src/coreclr/jit/lower.cpp

```diff
@@ -8202,6 +8309,49 @@ void Lowering::LowerStoreIndirCoalescing(GenTreeStoreInd* ind)
                 return;
         }
 
+        // For the memory move case we need to make sure pointers don't alias, e.g.:
+        //
+        //   *(a + 0) = *(b + 0);
+        //   *(a + 4) = *(b + 4);
+        //
+        // We can't merge this to
+        //
+        //   *(long*)(a + 0) = *(long*)(b + 0);
+        //
+        // if we don't know that a and b don't alias.
+        //
+        if (ind->Data()->OperIs(GT_IND))
+        {
+            if (!currData.baseAddr->TypeIs(TYP_REF) || !currValueData.baseAddr->TypeIs(TYP_REF))
+            {
+                // Non-TYP_REF pointers can alias,
+                // and we unlikely have enough information to prove otherwise.
+                return;
+            }
+
+            if (currData.index != nullptr || currValueData.index != nullptr)
+            {
+                // We only support base + constant offset, unknown index may lead to aliasing.
+                return;
+            }
+
+            // NOTE: for baseAddr and index we only checked curr* because prev* are guaranteed to be the same.
+
+            // Assume the worst case - currData.baseAddr and currValueData.baseAddr are the same GC object:
+            // Let's see if we can prove that they don't alias by checking the offsets.
+            const int storeStart = min(currData.offset, prevData.offset);
+            const int loadStart  = min(currValueData.offset, prevValueData.offset);
+
+            const int smallerOffset = min(storeStart, loadStart);
+            const int largerOffset  = max(storeStart, loadStart);
+
+            if ((smallerOffset != largerOffset) && ((smallerOffset + (int)genTypeSize(newType)) > largerOffset))
+            {
+                // May alias
+                return;
+            }
```

Comment:
Not sure I totally understand this logic. Can it use a normal interval intersection check? Like e.g.

https://github.com/dotnet/runtime/blob/c9088fe704ff2e3eab3e446ff5128843c5ea9cd8/src/coreclr/jit/promotion.cpp#L1510-L1533